src/lesson11-homework.ts

```diff
@@ -1,59 +1,174 @@
-        phone: "+3809625426549",
-        age: 30,
-        birthDate: "1993-02-13",
-      },
-    ];
+    return this.list.filter((user)=>{
+      if(user.phone){
+        const phoneNumbers = user.phone.match(/\d/g)?.join('')
+        if(phoneNumbers?.startsWith("380")){
+          return user
+        }
+      }
+    })
   }
 
   // TODO: Implement method that returns postal codes grouped by state, using this.list user data
-  getStatePostalCodes() {
-    return [
-      { name: "CO", postalCodes: ["80003", "80004"] },
-      { name: "AK", postalCodes: ["10001"] },
-    ];
+  getStatePostalCodes():TSimplifiedAddress[] {
+    const postalCodeUsers = this.list.filter((user)=>user.address?.postalCode && user.address.state)
+
+    const result:TSimplifiedAddress[] = []
+
+    for(const user of postalCodeUsers){
+      if(user.address){
+        const {state, postalCode} = user.address
+        const targetAddressIndex:number = result.findIndex((customAddress:TSimplifiedAddress)=>customAddress.name === state)
+
+        if(targetAddressIndex!==-1){
+          result[targetAddressIndex].postalCodes.push(postalCode)
+        } else{
+          const newCustomAddress:TSimplifiedAddress = {name:state, postalCodes:[postalCode]}
+          result.push(newCustomAddress)
+        }
+      }
+    }
+
+    return result
+  }
+
+  getAverageWomenAge(){
+    const womenArray = this.list.filter((user)=>{
+      if(user.gender === "female" && user.age){
+        return user
+      }
+    })
+    
+    const sumOfAllAges = womenArray.reduce((accumulator, woman)=>{
+      if(woman.age){
+        return accumulator + woman.age
+      }
+
+      return accumulator
+    }, 0)
+
+    return Math.floor(sumOfAllAges/womenArray.length)
+  }
+
+  getMostCommonWomanHairColor(){
+    const womenArray = this.list
+      .filter((user)=>user.gender === "female" && user.hair?.color)
+
+    const womenHairDictionary:TEntityCounterDictionary = {}
+
+    for(let i = 0; i<womenArray.length; i++){
+      const temporaryWomanCopy:TSimplifiedUser = {...womenArray[i]}
+      
+      if(temporaryWomanCopy.hair !== undefined){
+        if(!womenHairDictionary[temporaryWomanCopy.hair.color]){
+          womenHairDictionary[temporaryWomanCopy.hair.color] = 1
+        } else{
+          womenHairDictionary[temporaryWomanCopy.hair.color]++
+        }
+      }
+    }
+
+    return getMostCommonEntity(womenHairDictionary)
+  }
+
+
+  getMostCommonManBloodType(){
+     const menArray = this.list
+      .filter((user)=>user.gender === "male" && user.bloodGroup)
+
+    const menHairDictionary:TEntityCounterDictionary = {}
+
+    for(let i = 0; i<menArray.length; i++){
+      const temporaryManCopy:TSimplifiedUser = {...menArray[i]}
+      
+      if(temporaryManCopy.bloodGroup !== undefined){
+        if(!menHairDictionary[temporaryManCopy.bloodGroup]){
+          menHairDictionary[temporaryManCopy.bloodGroup] = 1
+        } else{
+          menHairDictionary[temporaryManCopy.bloodGroup]++
```

Comment:
[‣](https://github.com/Users/lyubomyrrudko32/development/typescript-course/lesson11-homework.html?grep=Users%20class%20getMostCommonManBloodType%20should%20work%20with%20real%20data)
AssertionError: expected 'Black' to deeply equal 'A-'
    at Context.<anonymous> (dist/lesson11-homework-test.js:313:65)